### PR TITLE
[WIP] Promise

### PR DESCRIPTION
**Pull request by rhysd**
In Vim8 or Neovim, some asynchronous operations are added like job, channel or timer. They're recommended because async operations don't block user input. However, we don't want to see nested callbacks like good old JavaScript. In ES6 or later, `Promise` is one of the answers to  callbacks hell.

APIs are almost the same as ECMAScript ([ECMA-262](https://www.ecma-international.org/publications/standards/Ecma-262.htm)) spec. And MDN is also a good documentation for Promise. ([Japanese](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Promise) or [English](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise))

Promise is a unidirectional state machine like following figure (pending state will move to fulfilled or rejected, but fulfilled or rejected states cannot move to pending):

fig

I ported ES6 promise to Vim script. I finished only library implementation. Tests and docs should be added. But I created this PR because I want to discuss some points before writing tests and docs. I referred [es6-promise](https://github.com/stefanpenner/es6-promise) a lot, which is the most popular JavaScript Promise polyfill.

```vim
let s:P = vital#vital#import('Promise')

" Use '.then' for chaining async operations
let p = s:P.new({resolve -> timer_start(1000, resolve)})
call p.then({-> execute('echom "1sec!"', '')})

" Use '.catch' for chaining async error handlings
let p = s:P.new({_, reject -> timer_start(500, {-> reject('Too late')})})
call p.catch({reason -> execute('echom "error! reason:" . reason', '')})

let waits = [1000, 500, 2000, 1500]

" Order of fulfillment: 1 -> 0 -> 3 -> 2
let wait_all = s:P.all(map(waits, {idx, w -> s:P.new({resolve -> timer_start(w, {-> resolve(idx)})})}))

" Output: [0, 1, 2, 3]
call wait_all.then({arr -> execute('echom string(arr)', '')})
```

Supported APIs are below:

- `s:Promise.new(resolver)` makes a new promise value (corresponding to `new Promise(resolver)` in JS.
- `s:Promise.resolve(val)`
- `s:Promise.reject(val)`
- `{promise value}.then(on_fulfilled[, on_rejected])`
- `{promise value}.catch(on_rejected)`
- `s:Promise.all(array)`
- `s:Promise.race(array)`
- `s:Promise.is_promise(val)` returns `val` is a promise or not as boolean
- `s:Promise.is_available()` returns whether this module is available or not in current environment

[bluebird](https://github.com/petkaantonov/bluebird) has more nice APIs. But I add only standard APIs in this PR.

Note that this library is currently only for Vim8 because internally lambda expressions are used. I need to rewrite them with `Partial` later. I'm considering to support Neovim and Vim8+.

[Japanese blogpost](http://rhysd.hatenablog.com/entry/2017/12/19/000810)

### 1. `s:Promise.new`

In JavaScript, `new Promise((resolve, reject) => ...)` works nicely. ~~However, in Vim script, we need to start argument with upper case as below because parameters of the lambda are funcref and not prefixed with `a:`.~~ Func ref arguments of lambda can start with lower case.

```vim
let p = s:Promise.new({resolve, reject -> resolve(42)})
```

I feel this does not look not so good. My idea is changing API of `s:Promise.new` as following.

```vim
let p = s:Promise.new({e -> e.resolve(42)})
```

Here, the `e` is an 'executor', which has fulfillment function and rejection function as its properties. (As bonus of this idea, I guess Vim7 can also be supported.) However, this interface is different from JS. It may be confusing for the people who knows JS's promise. What do you think?

### 2. Default result value of fulfillment

In JavaScript, default fulfilled result is `undefined` as below.

```javascript
console.log(Promise.resolve()) // output: undefined
```

However, Vim script does not have such a good(?) default value like that. Currently I use `v:null` for the purpose. However `0` may be more suitable because `:function`'s default value is `0`.

### 3. How to test async operations with vim-themis

This is just a question. How can I test async operations with vim-themis? I could not find an interface for them. For example, [mocha](https://mochajs.org/) can wait async results with `done` callback.

```javascript
it('does something', function(done) {
  setTimeout(done, 1000)
}
```

### 4. Module name

Currently I put this module as `Vital.Promise`. But it may not be appropriate. Candidates are below.

- `Vital.Promise` (for now)
- `Vital.Async.Promise` (may consider other asynchronous types such as observable)
- `Vital.Data.Promise` (but is promise a data structure? 'data structure' is so fuzzy word...)

------

Both English and Japanese comments are welcome! Give me your thoughts.

### CODE REVIEW

**Comment by rhysd** [comment]
コードレビューはまだできるレベルになってませんが，上記3点相談させてください！

File: autoload/vital/__vital__/Promise.vim [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+function! s:_handle_thenable(promise, thenable) abort
+  if thenable._state == s:FULFILLED
+    call s:fulfill(a:promise, a:thenable._result)
+  elseif thenable._state == s:REJECTED
+    call s:reject(a:promise, a:thenable._result)
+  else
+    call s:_subscribe(
+          \ a:thenable,
+          \ v:null,
+          \ {Val -> s:_resolve(a:promise, Val)},
```
Comment (rhysd) [inline autoload/vital/__vital__/Promise.vim]: :(

**Comment by thinca** [comment]
1.

> ```vim
> let p = s:Promise.new({e -> e.resolve(42)})
> ```

👍 This also has extensibility.

2.

Or... should not we omit the result value?  Is it inconvenient?
(This is difficult problem.)

3.

> How can I test async operations with vim-themis?

Unfortunately, there is no way to do that.  I want to support it.
BTW, latest version of mocha can async test by returning the Promise object.  I like this.
https://mochajs.org/#working-with-promises

**Comment by rhysd** [comment]
@thinca 

コメントありがとうございます．

> Or... should not we omit the result value? Is it inconvenient?

はい，デフォルトで `_result` キーがセットされていないと，`_result` の値を参照する時に毎回 `has_key` でチェックしてキーがあるときとないときで別処理を書く必要があります．

> BTW, latest version of mocha can async test by returning the Promise object. I like this.

超同意なんですが，それにはこの PR が入る必要があり，デッドロック… 多分 Promise のテストのために Promise に依存した機能を使うのはテストが落ちた時に切り分けが面倒になりそうです．

**Comment by tyru** [comment]
@rhysd ++

1.

```viml
let p = s:Promise.new({e -> e.resolve(42)})
```

自分もこっちのがいいと思います。理由は以下2点です。

* ES Promise で現状サポートされてない cancel が入るかもしれないという話もある
  * 「元々のPromise仕様では、ある値の非同期の読み込みが進行しているときに、それをキャンセルすることの意味は考慮されていませんでした」 via [POSTD](http://postd.cc/async-and-await/) and [Presentation](https://docs.google.com/presentation/d/1V4vmC54gJkwAss1nfEt9ywc-QOVOfleRxD5qtpMpc8U/edit#slide=id.gc6f9e470d_0_0)
* 引数が増えた時 JS と違って Vim script では引数の数が正確に合ってないとエラーが出る

あとちなみにですが、

> we need to start argument with upper case

その必要はないはずです。
`a:` 変数はもともと Funcref であっても upper case である必要がなく、
加えて lambda の引数は `a:` を省略できますが、内部的には `a:` を[補間してくれてるだけみたいです](https://twitter.com/k_takata/status/873391546233532420)（実際今以下を試して大丈夫でした）。

```viml
function! Succ(value)
  return a:value + 1
endfunction
" echo 2
echo {resolve,reject -> resolve(1)}(function('Succ'), function('Succ'))
```

**Comment by rhysd** [comment]
@tyru 

> ES Promise で現状サポートされてない cancel が入るかもしれないという話もある

確かに．cancelable promise の話は確か ES Promise では実装しないとなったと思いますが，Vim script では実装するという選択肢はありますね．@thinca さんのおっしゃるように拡張性を担保する価値がありそうです．

> 引数が増えた時 JS と違って Vim script では引数の数が正確に合ってないとエラーが出る

これはラムダ式については若干違います．例を拝借して少しいじると…

```vim
function! Succ(value)
  return a:value + 1
endfunction
" echo 2
echo {resolve -> resolve(1)}(function('Succ'), function('Succ'))
```

これでもエラーは出ません．どうやら引数が少ない分にはエラーが出ません（ただ，どうやら undocumented なので依存しないほうが良いかもしれません）

> その必要はないはずです。
> a: 変数はもともと Funcref であっても upper case である必要がなく、
> 加えて lambda の引数は a: を省略できますが、内部的には a: を補間してくれてるだけみたいです（実際今以下を試して大丈夫でした）。

あれ，おっしゃる通りでした… 僕が勘違いしていました．ありがとうございます．ただしこれは内部実装に依存した話でドキュメントにはざっと見た感じ明記されていない気がしました．

**Comment by rhysd** [comment]
議論したい項目に 4. モジュール名を追加しました．個人的には `Vital.Async.Promise` が良いのかな？と思うのですが，どうでしょう？

**Comment by tyru** [comment]
`Async.Promise` でいいと思います。

File: autoload/vital/__vital__/Promise.vim (hunk 1 above)
Comment (mattn) [inline autoload/vital/__vital__/Promise.vim]: oh

**Comment by thinca** [comment]
`Async.***` になりそうな他のモジュールの候補って現時点で何か想定ありますか?

**Comment by rhysd** [comment]
一応脳内で想定していたのは `Async.Observable` です．こないだ @tyru さんがそんな感じのことをおっしゃっていたので…

File: autoload/vital/__vital__/Promise.vim [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+let s:DICT_T = type({})
+let s:NULL_T = type(v:null)
+
+function! s:noop() abort
+endfunction
+let s:NOOP = function('s:noop')
+
+" Internal APIs
+
+" TODO: v:null or 0, which should the default result value be ?
```
Comment (lambdalisue) [inline autoload/vital/__vital__/Promise.vim]: :+1: for `v:null`

**Comment by rhysd** [comment]
質問項目の 1. について考えてみたんですが，やはり `{resolve, reject -> expr}` のほうが `{exe -> expr}` より良いかなと思っています．

`.then` の引数は `.then(resolve, reject)` で，そことの対応が崩れてしまうのと，ES6 の Promise のインターフェースからやはりなるべく外したくないというのが意図です．

Vim script のラムダ式の引数は `...` が暗黙的に引数末尾に追加されるというのを知ったので，`.new()` するときにも `reject` は書かなくて良い（`s:P.new({resolve -> ... }`）のと，ラムダ式を使わない場合でも，

```vim
function s:resolveFoo(resolve, ...)
  " do something ...
  let foo = ...
  resolve(foo)
endfunction
```

みたいな感じで `...` を書いてもらえれば良いかなと．さらに，このプルリクをつくった時点では neovim にラムダ式が入っていなかったのでそこも懸念点でしたが，今は普通に使えるようですし．

**Comment by tyru** [comment]
> Vim script のラムダ式の引数は ... が暗黙的に引数末尾に追加される

知らなかった…
`.then(resolve, reject)` :+1:

File: autoload/vital/__vital__/Async/Promise.vim [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+endfunction
+
+function! s:PROMISE.then(...) abort dict
+  let parent = self
+  let state = parent._state
+  let child = s:new(s:NOOP)
+  if state == s:FULFILLED
+    " In ECMAScript spec, this callback must be called asynchronously, but Vim script does not have
+    " asynchronous function such as setTimeout(). So call the callback synchronously here.
+    call s:_invoke_callback(state, child, get(a:000, 0, v:null), parent._result)
```
Comment (lambdalisue) [inline autoload/vital/__vital__/Async/Promise.vim]: ここ timer_start ではダメですか？
メインスレッドがビジー状態では無くなった時に実行されるという意味で setTimeout  とほぼ同義なので

**Comment by mattn** [comment]
In my point view, LGTM. :+1:  leave for others. :hand:

File: autoload/vital/__vital__/Async/Promise.vim (hunk 3 above)
Comment (rhysd) [reply autoload/vital/__vital__/Async/Promise.vim]: ブログ記事にも書きましたが，`.then` のコールバック自体は今は同期的に実行するようになっています．確かにここは ES6 Promise とは異なる挙動になっているのですが，`.then` のたびにタイマーを挟む（例えば `.race()` だと中で引数のリストの要素数分だけ `.then` が呼ばれます）のは（最適化されまくっている JavaScript に比べて）Vim script ではオーバーヘッドが大きいかなと思って今はこうなっています．いずれベンチマーク取ってみて確認が必要かなと思ってますが，同期的に実行して困るケースが思いつかなかったので今はそのままになっています．何か困るケースってありますか？

**Comment by lambdalisue** [comment]
例えば一つの then に 10ms かかると仮定します。
race で 100 個の promise を同期的に処理すると 1s かかってしまいユーザビリティに影響が出ると思います。一方でタイマーを使うと 10ms 毎に分割された処理が100回非ビジーの際に走るためユーザビリティを損ないにくいと思います。

上記に関しては書いてみたわけではないので確定ではないです。ただ gina にてコマンドの実行結果をバッファに書き込む際に小さなチャンクに分けてタイマーで定期的に書き込むというようなことをしています。これによりロード中でもカーソルの引っ掛かりが低減したので、上記のような問題があるかなと推測しています。

**Comment by lambdalisue** [comment]
race じゃやくて all のが例としては良さそうですね

**Comment by rhysd** [comment]
はい，ポイントは私も理解しているつもりで，コメントに書いた通りベンチマークを取ってみたいと考えています．

> 上記に関しては書いてみたわけではないので確定ではないです。ただ gina にてコマンドの実行結果をバッファに書き込む際に小さなチャンクに分けてタイマーで定期的に書き込むというようなことをしています。これによりロード中でもカーソルの引っ掛かりが低減したので、上記のような問題があるかなと推測しています。

具体例ありがとうございます．これが知りたかったので助かります（要は実践的に困る場合があるのかという点）．検討してみます．

**Comment by lambdalisue** [comment]
:+1:

**Comment by rhysd** [comment]
https://github.com/vim-jp/vital.vim/pull/526#discussion_r157657536 について，試してみましたが，現状だと23回 `.then` がネストすると `maxfuncdepth`（既定値は100）に当ってしまいますね…原因は詳しくまだ見てませんが，下記に試したことをとりあえずまとめておきます．

まずは下記のパッチを当てます．`a:000` のキャプチャはラムダ式の実行が非同期になるとうまく動かなかった（変な値になっていた）ので，予め `Res` と `Rej` という変数を用意することにしたら動きました．

```diff
diff --git a/autoload/vital/__vital__/Async/Promise.vim b/autoload/vital/__vital__/Async/Promise.vim
index d7f952f0..705fee29 100644
--- a/autoload/vital/__vital__/Async/Promise.vim
+++ b/autoload/vital/__vital__/Async/Promise.vim
@@ -233,16 +233,14 @@ function! s:_promise_then(...) dict abort
   let parent = self
   let state = parent._state
   let child = s:new(s:NOOP)
+  let Res = get(a:000, 0, v:null)
+  let Rej = get(a:000, 1, v:null)
   if state == s:FULFILLED
-    " In ECMAScript spec, this callback must be called asynchronously, but Vim script does not have
-    " asynchronous function such as setTimeout(). So call the callback synchronously here.
-    call s:_invoke_callback(state, child, get(a:000, 0, v:null), parent._result)
+    call timer_start(0, {-> s:_invoke_callback(state, child, Res, parent._result)})
   elseif state == s:REJECTED
-    " In ECMAScript spec, this callback must be called asynchronously, but Vim script does not have
-    " asynchronous function such as setTimeout(). So call the callback synchronously here.
-    call s:_invoke_callback(state, child, get(a:000, 1, v:null), parent._result)
+    call timer_start(0, {-> s:_invoke_callback(state, child, Rej, parent._result)})
   else
-    call s:_subscribe(parent, child, get(a:000, 0, v:null), get(a:000, 1, v:null))
+    call s:_subscribe(parent, child, Res, Rej)
   endif
   return child
 endfunction
```

試したスクリプト:

```vim
let s:P = vital#vital#import('Async.Promise')
let s:times = range(25)

let p = s:P.new({resolve -> resolve(0)})
for s:i in s:times
    let p = p.then({i -> i + 1})
endfor

call p.then({i -> execute('echomsg ' . string('Done: ' . i), '')})
     \.catch({e -> execute('echomsg ' . string('ERROR: ' . i), '')})
     \.then({-> execute('exit')})
```

`s:times` を `range(23)` 以上にすると `maxfuncdepth` のエラー E132 が出てしまいます

```
Error detected while processing function <lambda>237[1]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SN
R>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__
fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback
[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__
publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolv
e[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201_
_invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfil
l[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..
<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publis
h[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..
<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invok
e_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..
<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>2
01__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15].
.<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>2
01__fulfill[9]..<SNR>201__publish[15]..<SNR>201__invoke_callback[17]..<SNR>201__resolve[4]..<SNR>201__fulfill:
line    9:
E132: Function call depth is higher than 'maxfuncdepth'
```

つまり，`.then` や `.catch` は22までしか chain できない（かつ，コールバックの中で関数呼出しがある場合さらに減る？）です．内部的に `.then` を呼んでいる `.race` の引数のリストの要素数も22までになってしまいます．

`.then` 内で `timer_start` を使わない今の実装であれば `let p = p.then(...)` が同期的に実行されるのでこの問題は起きていないです（`s:times` を 10000 とかにしても ok）

**Comment by lambdalisue** [comment]
`maxfuncdepth` 問題は回避が出来ないので入れ子にしすぎるなしか解決策がない気がします。

ちなみに現在の `rhysd/vital.vim@promise` のコードにて、以下のように `resolve` が非同期的に解決されると 23 回で同じ問題が発生しました。

```vim
let s:P = vital#vital#import('Async.Promise')
let s:times = range(23)

let p = s:P.new({resolve -> timer_start(1000, resolve)})
for s:i in s:times
    let p = p.then({i -> i + 1})
endfor

call p.then({i -> execute('echomsg ' . string('Done: ' . i), '')})
     \.catch({e -> execute('echomsg ' . string('ERROR: ' . i), '')})
```

```
"~/Code/github.com/vim-jp/vital.vim/test.vim" 11L, 376C
Done: 38
Error detected while processing function <lambda>4[1]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish[15]..<SNR>276__invoke_callback[17]..<SNR>276__resolve[4]..<SNR>276__fulfill[9]..<SNR>276__publish:
line   15:
E132: Function call depth is higher than 'maxfuncdepth'
```

あと `Res` や `Rej` を使わなくても`s:_invoke_callback()` の定義に `...` を足すことで `timer_start(0, function('s:_invoke_callback', [state, child, get(a:000, 0, v:null), parent._result]))` と直接呼べます。

**Comment by rhysd** [comment]
確かに `...` 足すと `a:` 使えました（別の箇所ですでにやっていたのに忘れてた…）．

そして，下記だと問題ないことが分かりました．うーん，正直タイマーの実装を分かっていないのが厳しい…

```vim
let s:P = vital#vital#import('Async.Promise')
let s:times = range(100)

function! s:timer(ms) abort
    return s:P.new({res -> timer_start(a:ms, res)})
endfunction

let p = s:P.new({resolve -> resolve(0)})
for s:i in s:times
    let p = p.then({-> s:timer(10)})
endfor

call p.then({i -> execute('echomsg ' . string('Done: ' . i), '')})
     \.catch({e -> execute('echomsg ' . string('ERROR: ' . i), '')})
```

**Comment by rhysd** [comment]
すみません，実装ミスってました．下記なら問題なく動きますね :tada:

```diff
diff --git a/autoload/vital/__vital__/Async/Promise.vim b/autoload/vital/__vital__/Async/Promise.vim
index d7f952f0..f2db5670 100644
--- a/autoload/vital/__vital__/Async/Promise.vim
+++ b/autoload/vital/__vital__/Async/Promise.vim
@@ -129,9 +129,7 @@ function! s:_fulfill(promise, value) abort
   let a:promise._result = a:value
   let a:promise._state = s:FULFILLED
   if !empty(a:promise._children)
-    " In ECMAScript spec, this callback must be called asynchronously but Vim script does not have
-    " asynchronous function such as setTimeout(). So call the callback synchronously here.
-    call s:_publish(a:promise)
+    call timer_start(0, {-> s:_publish(a:promise)})
   endif
 endfunction
 
@@ -141,9 +139,7 @@ function! s:_reject(promise, reason) abort
   endif
   let a:promise._result = a:reason
   let a:promise._state = s:REJECTED
-  " In ECMAScript spec, this callback must be called asynchronously, but Vim script does not have
-  " asynchronous function such as setTimeout(). So call the callback synchronously here.
-  call s:_publish(a:promise)
+  call timer_start(0, {-> s:_publish(a:promise)})
 endfunction
 
 function! s:_resolve_one(index, value) dict abort
@@ -233,16 +229,14 @@ function! s:_promise_then(...) dict abort
   let parent = self
   let state = parent._state
   let child = s:new(s:NOOP)
+  let Res = get(a:000, 0, v:null)
+  let Rej = get(a:000, 1, v:null)
   if state == s:FULFILLED
-    " In ECMAScript spec, this callback must be called asynchronously, but Vim script does not have
-    " asynchronous function such as setTimeout(). So call the callback synchronously here.
-    call s:_invoke_callback(state, child, get(a:000, 0, v:null), parent._result)
+    call timer_start(0, {-> s:_invoke_callback(state, child, Res, parent._result)})
   elseif state == s:REJECTED
-    " In ECMAScript spec, this callback must be called asynchronously, but Vim script does not have
-    " asynchronous function such as setTimeout(). So call the callback synchronously here.
-    call s:_invoke_callback(state, child, get(a:000, 1, v:null), parent._result)
+    call timer_start(0, {-> s:_invoke_callback(state, child, Rej, parent._result)})
   else
-    call s:_subscribe(parent, child, get(a:000, 0, v:null), get(a:000, 1, v:null))
+    call s:_subscribe(parent, child, Res, Rej)
   endif
   return child
 endfunction
```

**Comment by lambdalisue** [comment]
timer 引数握りつぶす為だけにラムダを使うよりは、直接 function で引数を固定させた方が良いかなと思いました。ちょっとこの辺の実装は分かっていないのですが、ラムダでラップするとその分スタック消費する気がします。

**Comment by lambdalisue** [comment]
Vim の maxdepth 難しい。

話は違うかもしれませんが gina だと job の callback の中で sleep 1m を呼ぶと、その部分から別のコールバックが呼ばれたように扱われてしまい maxdepth に到達したことがあります（sleep で処理が別に渡るので、その部分から job の別のコールバックが呼ばれる形になって再起になるから maxdepth に到達する）

上記があるのでもしかすると then の中で sleep 呼ぶと同じ現象になるかもしれません。

**Comment by rhysd** [comment]
なるほど partial 使ったほうが良さげかもですね．下記の順序で修正してみます:

1. 既存実装と `timer_start` 噛ました版でどれくらいパフォーマンスが違うか確認
2. テスト修正
3. ラムダ式ではなく partial 使うようにする

ちなみに上記実装では funcmaxdepth の問題は（指摘いただいている `:sleep` の話以外は）解消しているようにみえるので，コールスタックの深さは気にしなくて良いように思います．

**Comment by lambdalisue** [comment]
ありがとうございますー
他の部分のレビューはその後の方が良いでしょうか？

**Comment by rhysd** [comment]
はい、WIP 外れたらでお願いします（いくつか修正すべき点があるのは認識済みです）

**Comment by rhysd** [comment]
ざっくりベンチマーク取ってみたのですが、`timer_start` が2回噛むほうは14倍ぐらい遅いですね

- `timer_start` なし同期実行版 `.then` で10000回 `.then` をつなげてみる
```
vim -X -N -u NONE -i NONE -V1 -e -s -S run_bench.vim  4.13s user 0.48s system 26% cpu 17.689 total
```

- `timer_start` あり非同期実行版 `.then` で10000回 `.then` をつなげてみる
```
vim -X -N -u NONE -i NONE -V1 -e -s -S run_bench.vim  22.28s user 0.03s system 85% cpu 26.054 total
```

partial 使った版も見てみましたがほぼ変わりませんでした（タイマー発火部分が支配的っぽい）

**Comment by lambdalisue** [comment]
> はい、WIP 外れたらでお願いします

フライングすみません :bow: 💦

**Comment by lambdalisue** [comment]
個人的には非同期の時と同期の時とで、ベンチマークを比較することにあんまり意味ないかなと思います（メインスレッドの忙しさで14倍以上になりえるので）。

どちらかというと then のチェインが実行されている間でもカーソルが動かせる等、メインスレッドを占有していないか？の方が今回の場合は重要な気がします（そして、それを定量的に測るのは難しすぎる）

**Comment by lambdalisue** [comment]
↑ にしても 14 倍はちょっと‥‥と思うところもありますねw

**Comment by rhysd** [comment]
試しに

```vim
let s:p = s:P.resolve()
while 1
    let s:p = s:p.then({-> v:null})
endwhile
```

みたいな無限ループ起こすスクリプトを `:source` してみたところユーザの入力は完全にブロッキングしてしまいますね…help のテキストを読む感じ，飽くまで Vim が入力可能になるまで待つだけで，ユーザの入力を優先するようには動いていないようにみえます．

**Comment by rhysd** [comment]
確認してみたところ，`:so` によるスクリプトの評価が終わってからはユーザ入力はスムーズに行えますね．なので，`.then` のチェーンの発火を待っている間にユーザの入力を阻害することはないみたいです．

File: autoload/vital/__vital__/Async/Promise.vim [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+  else
+    let Result = a:result
+  endif
+
+  if a:promise._state != s:PENDING
+    " Do nothing
+  elseif has_callback && success
+    call s:_resolve(a:promise, Result)
+  elseif !success
+    call s:_reject(a:promise, Err)
```
[the remark this answers is not in the document]
Comment (rhysd) [reply autoload/vital/__vital__/Async/Promise.vim]: 何回言うねん

**Comment by rhysd** [comment]
I created another PR #567 to review this library because this PR is too long. Please leave review comments there.

**Pull request closed without merging**